# TensorFlow

TensorFlow is a library developed by the team of Google that can be used to program various machine learning/deep learning algorithms. Tensorflow has its own way to define datatypes, which we must learn in order to program various algorithms. 

`Data Flow Graph` is an important feature of Tensorflow. This concept defines the operation of a program as a `graph`, where each basic mathematical oepration is a _node_ and the data that undergoes the mathematical operation is a _edge_. More generally, the edges or the datatypes are multidimensional tensors. 

Following is the version of TensorFlow where this tuorial has been tested. 

In [1]:
import tensorflow as tf
print(tf.__version__)

2.8.0


## Start a session and work with constants and variables

Let's look at an example. Let's write a code to convert an angle from degrees to radians using standard Python3 and write the same code using tensor flow.

In [2]:
import numpy as np

ang2rad=np.pi/180.0
angle_degrees=45.0
theta=angle_degrees*ang2rad

print(theta)

0.7853981633974483


Now let's learn how write the same code using tensorflow. We will first declare a constant and see how to print it.  In Python3, what we would do is the following.

In [3]:
pi=np.pi

print(pi)

3.141592653589793


In TensorFlow, you have to define all your data as a constant or a variable. So, we can try the following code.

In [4]:
pi=tf.constant(np.pi)

print(pi)

tf.Tensor(3.1415927, shape=(), dtype=float32)


Now you can see that the constant is stored as a Tensor (with the shape of a scalar) but the output is not expected. Suppose we only want to see the number `3.1415927` as the output, then we have work with a `session`.

In [5]:
with tf.compat.v1.Session() as sess:
    pi=tf.constant(np.pi)
    result = sess.run(pi)

    print(result)

3.1415927


You must have noticed that our result using standard Python has more significant figure, `3.141592653589793`, than the result printed using TensorFlow, `3.1415927`. We can provide another argument `dtype` as we have done for `numpy` and state that we are interested in a 64-bit float.

In [6]:
with tf.compat.v1.Session() as sess:
    pi=tf.constant(np.pi,dtype=tf.float64)
    result = sess.run(pi)

    print(result)

3.141592653589793


Now, let's move to writing the entire code to convert the angle from degrees to radian

In [7]:
# Start a `Session` to execute the graph.
with tf.compat.v1.Session() as sess:

    # Build a dataflow graph.
    ang2rad = tf.constant(np.pi/180.0,dtype=tf.float64)       # Edge
    angle_degrees= tf.constant(45.0,dtype=tf.float64)         # Edge
    theta=tf.multiply(angle_degrees,ang2rad)                  # Node connecting the 2 edges

    # Execute the graph and store the output
    result = sess.run(theta)
    print(result)

0.7853981633974483


Yes, the code is aready getting a little lengthy. But, still let's try to declare `angle_degrees` as a variable because, in particle, this is how we might use this piece of code. For every angle in degrees, we may want a value in radian.

In [8]:
# Start a `Session` to execute the graph.
with tf.compat.v1.Session() as sess:

    # Build a dataflow graph.
    ang2rad = tf.constant(np.pi/180.0,dtype=tf.float64)                   # Edge-constant
    
    angle_degrees= tf.Variable(initial_value=45.0,dtype=tf.float64)       # Edge-variable
    
    # We have to create a session to initialize the variable
    init = tf.compat.v1.global_variables_initializer()        
    sess.run(init)

    theta=tf.multiply(angle_degrees,ang2rad)                             # Node connecting the 2 edges

    # Execute the graph and store the output
    result = sess.run(theta)
    print(result)

0.7853981633974483


Now, we can write a function and run a TensorFlow session in it and use it in Python.

In [9]:
def tf_deg2rad(angle):
    import tensorflow as tf
    
    with tf.compat.v1.Session() as sess:
    
        ang2rad = tf.constant(np.pi/180.0,dtype=tf.float64) # node-1
        
        # declare a variable and initialize (i.e. run) it
        angle_degrees= tf.Variable(initial_value=angle,dtype=tf.float64)  # node-2    
        init = tf.compat.v1.global_variables_initializer()        
        sess.run(init) 

        # define a graph and run it
        theta=tf.multiply(angle_degrees,ang2rad)        # graph     
        result = sess.run(theta)
        
    return result

In [10]:
for angle in [30,45,60,90,180]:
    print(tf_deg2rad(angle))

0.5235987755982988
0.7853981633974483
1.0471975511965976
1.5707963267948966
3.141592653589793
